updates for iosxe, nxos show platform and show interface, fix show interface with single ip helper address

mgdransfield opened the pull request:
a few parser updates for iosxe and nxos

tahigash commented on src/genie/libs/parser/iosxe/c9500/show_platform.py:
```diff
@@ -59,8 +59,12 @@ class ShowVersionSchema(MetaParser):
                     },
-                    'license_level': str,
-                    'next_reload_license_level': str,
+                    Optional('license_level'): {
```
I think we don't need to change these because both are corresponding to below.
```
        AIR License Level: AIR DNA Advantage

        Next reload AIR license Level: AIR DNA Advantage

tahigash replied:
added new comment for this with suggested codes.

tahigash commented on src/genie/libs/parser/iosxe/show_interface.py:
```diff
@@ -255,9 +255,16 @@ def cli(self,interface="",output=None):
         # MTU 1500 bytes, BW 10000 Kbit, DLY 1000 usec, 
         # MTU 1600 bytes, sub MTU 1600, BW 3584 Kbit/sec, DLY 410 usec,
+        # MTU 1500 bytes, BW 5200 Kbit/sec, RxBW 25000 Kbit/sec, DLY 100 usec, 
+        p6 = re.compile(r'^MTU +(?P<mtu>\d+) +bytes(, +sub +MTU +'
```
looks good

tahigash commented on src/genie/libs/parser/iosxe/show_interface.py:
```diff
@@ -255,9 +255,16 @@ def cli(self,interface="",output=None):
+                        r'(?P<sub_mtu>\d+))?, +BW +(?P<bandwidth>[0-9]+) +Kbit(\/sec)?'
+                        r'(, +RxBW +[0-9]+ +Kbit(\/sec)?)?, +'
+                        r'DLY +(?P<delay>[0-9]+) +usec,$')
+        '''
```
please remove if this regex is not needed

tahigash replied:
confirmed

tahigash commented on src/genie/libs/parser/nxos/show_platform.py:
```diff
@@ -248,6 +248,22 @@ def cli(self,output=None):
+                continue
+
+            # cisco N9K-C93180YC-EX ("supervisor")
+            p15a = re.compile(r'^\s*cisco +(?P<model>[a-zA-Z0-9\-\s]+)'
```
other than creating new regex, I assume we can change p15 regex to like below?
```
^\s*cisco +(?P<model>[a-zA-Z0-9\-\s]+)( +\((?P<slot>[0-9]+) Slot\))?( +[C|c]hassis)?( +\(\"(?P<rp>[a-zA-Z0-9\s\-\/\+]+)\"\))?(\s)?$

tahigash replied:
still seeing p15a... not updated yet?

tahigash replied:
confirmed

mgdransfield commented:
I have updated and pushed again.
The first change is required when using smart licensing so I have left it
in.
The nxos change fails tests using your suggestion but I updated it to
remove the ? after chassis and it seems ok
other items tidied up

Regards,
Mike

On Tue, Aug 11, 2020 at 9:42 AM Takashi Higashimura <
notifications@github.com> wrote:

> *@tahigash* requested changes on this pull request.
> ------------------------------
>
> In src/genie/libs/parser/iosxe/c9500/show_platform.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#discussion_r468238690>
> :
>
> > @@ -59,8 +59,12 @@ class ShowVersionSchema(MetaParser):
>                      Optional('mem_size'): {
>                          Any(): str,
>                      },
> -                    'license_level': str,
> -                    'next_reload_license_level': str,
> +                    Optional('license_level'): {
>
> I think we don't need to change these because both are corresponding to
> below.
>
>         AIR License Level: AIR DNA Advantage
>
>         Next reload AIR license Level: AIR DNA Advantage
>
> ------------------------------
>
> In src/genie/libs/parser/iosxe/show_interface.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#discussion_r468241559>
> :
>
> > @@ -255,9 +255,16 @@ def cli(self,interface="",output=None):
>          # MTU 1500 bytes, BW 768 Kbit/sec, DLY 3330 usec,
>          # MTU 1500 bytes, BW 10000 Kbit, DLY 1000 usec,
>          # MTU 1600 bytes, sub MTU 1600, BW 3584 Kbit/sec, DLY 410 usec,
> +        # MTU 1500 bytes, BW 5200 Kbit/sec, RxBW 25000 Kbit/sec, DLY 100 usec,
> +        p6 = re.compile(r'^MTU +(?P<mtu>\d+) +bytes(, +sub +MTU +'
>
> looks good
> ------------------------------
>
> In src/genie/libs/parser/iosxe/show_interface.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#discussion_r468241792>
> :
>
> > @@ -255,9 +255,16 @@ def cli(self,interface="",output=None):
>          # MTU 1500 bytes, BW 768 Kbit/sec, DLY 3330 usec,
>          # MTU 1500 bytes, BW 10000 Kbit, DLY 1000 usec,
>          # MTU 1600 bytes, sub MTU 1600, BW 3584 Kbit/sec, DLY 410 usec,
> +        # MTU 1500 bytes, BW 5200 Kbit/sec, RxBW 25000 Kbit/sec, DLY 100 usec,
> +        p6 = re.compile(r'^MTU +(?P<mtu>\d+) +bytes(, +sub +MTU +'
> +                        r'(?P<sub_mtu>\d+))?, +BW +(?P<bandwidth>[0-9]+) +Kbit(\/sec)?'
> +                        r'(, +RxBW +[0-9]+ +Kbit(\/sec)?)?, +'
> +                        r'DLY +(?P<delay>[0-9]+) +usec,$')
> +        '''
>
> please remove if this regex is not needed
> ------------------------------
>
> In src/genie/libs/parser/nxos/show_platform.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#discussion_r468244610>
> :
>
> > @@ -248,6 +248,22 @@ def cli(self,output=None):
>
>                  if 'rp' not in version_dict['platform']['hardware']:
>                      version_dict['platform']['hardware']['rp'] = rp
> +                continue
> +
> +            # cisco N9K-C93180YC-EX ("supervisor")
> +            p15a = re.compile(r'^\s*cisco +(?P<model>[a-zA-Z0-9\-\s]+)'
>
> other than creating new regex, I assume we can change p15 regex to like
> below?
>
> ^\s*cisco +(?P<model>[a-zA-Z0-9\-\s]+)( +\((?P<slot>[0-9]+) Slot\))?( +[C|c]hassis)?( +\(\"(?P<rp>[a-zA-Z0-9\s\-\/\+]+)\"\))?(\s)?$
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#pullrequestreview-464651139>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFM7PB7ZIIUJVQ6L42K7VT3SACAU5ANCNFSM4PRIQNZA>
> .
>

tahigash commented on src/genie/libs/parser/iosxe/c9500/show_platform.py:
```diff
@@ -55,12 +55,18 @@ class ShowVersionSchema(MetaParser):
                     },
-                    'license_level': str,
-                    'next_reload_license_level': str,
+                    Optional('license_level'): {
+                        Any(): str,
+                    },
+                    Optional('next_reload_license_level'): {
+                        Any(): str,
+                    },
```
```suggestion
                    Optional('license_level'): str,
                    Optional('next_reload_license_level'): str,
```
what I meant on above comment is this. do we have a reason why we need to have `Any()` under these levels?

tahigash replied:
confirmed

mgdransfield commented:
updated license level as suggested. Had another go at removing 15a and this time passes tests

tahigash commented on src/genie/libs/parser/iosxe/c9500/show_platform.py:
```diff
@@ -55,12 +55,14 @@ class ShowVersionSchema(MetaParser):
                     'mb_rev_num': str,
                     'model_num': str,
-                    'system_sn': str,
+                    Optional ('system_sn'): {
```
sorry, why change is needed? Please share sample output of `show version`

tahigash replied:
thank you for the output. I confirmed no SN in output.
```
System Serial Number               :
```
but this doesn't require to change the schema. let's change to `Optional` without having new level.
```
Optional('system_sn'): str,

tahigash replied:
confirmed

tahigash commented on src/genie/libs/parser/iosxe/show_interface.py:
```diff
@@ -2069,8 +2071,12 @@ def cli(self,interface="",output=None):
                     interface_dict[interface]['helper_address'] = \
-                        m.groupdict()['address']
+                        helper_list
+                    #interface_dict[interface]['helper_address'] = \
+                    #    m.groupdict()['address']
```
Please remove

tahigash replied:
confirmed the change

tahigash commented on src/genie/libs/parser/iosxe/tests/test_show_interface.py:
```diff
@@ -17398,6 +17398,257 @@ class TestShowIpInterface(unittest.TestCase):
         IPv4 WCCP Redirect exclude is disabled
     '''}
 
+    golden_parsed_output4 = {
```
Unittest style has been changed. now we have command output separately. 
Please find the detail from below page.
https://pubhub.devnetcloud.com/media/pyats-development-guide/docs/writeparser/writeparser.html#test-your-new-parser

`ShowIpInterface` parser class's test is now under below location.
https://wwwin-github.cisco.com/pyATS/genieparser/tree/dev/src/genie/libs/parser/iosxe/tests/ShowIpInterface/cli

Could you please adjust to above new style accordingly? Let us know if any questions.

tahigash replied:
confirmed the change

tahigash reviewed (changes requested):
added some comments. please check and let us know if any queries.

mgdransfield commented:
pushed changes to remove commented lines and updated test as requested.
Sample show version for 9500 attached

Mike

On Sat, Oct 3, 2020 at 6:08 AM Takashi Higashimura <notifications@github.com>
wrote:

> *@tahigash* requested changes on this pull request.
>
> added some comments. please check and let us know if any queries.
> ------------------------------
>
> In src/genie/libs/parser/iosxe/c9500/show_platform.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#discussion_r499024842>
> :
>
> > @@ -55,12 +55,14 @@ class ShowVersionSchema(MetaParser):
>                      'model_rev_num': str,
>                      'mb_rev_num': str,
>                      'model_num': str,
> -                    'system_sn': str,
> +                    Optional ('system_sn'): {
>
> sorry, why change is needed? Please share sample output of show version
> ------------------------------
>
> In src/genie/libs/parser/iosxe/show_interface.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#discussion_r499025269>
> :
>
> > +                    #interface_dict[interface]['helper_address'] = \
> +                    #    m.groupdict()['address']
>
> Please remove
> ------------------------------
>
> In src/genie/libs/parser/iosxe/tests/test_show_interface.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#discussion_r499026983>
> :
>
> > @@ -17398,6 +17398,257 @@ class TestShowIpInterface(unittest.TestCase):
>          IPv4 WCCP Redirect exclude is disabled
>      '''}
>
> +    golden_parsed_output4 = {
>
> Unittest style has been changed. now we have command output separately.
> Please find the detail from below page.
>
> https://pubhub.devnetcloud.com/media/pyats-development-guide/docs/writeparser/writeparser.html#test-your-new-parser
>
> ShowIpInterface parser class's test is now under below location.
>
> https://wwwin-github.cisco.com/pyATS/genieparser/tree/dev/src/genie/libs/parser/iosxe/tests/ShowIpInterface/cli
>
> Could you please adjust to above new style accordingly? Let us know if any
> questions.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#pullrequestreview-501413091>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFM7PB25FJBBMFG5A5XCLXLSIYXLRANCNFSM4PRIQNZA>
> .
>

Cisco IOS XE Software, Version 16.09.05
Cisco IOS Software [Fuji], Catalyst L3 Switch Software (CAT9K_IOSXE), Version 16.9.5, RELEASE SOFTWARE (fc1)
Technical Support: http://www.cisco.com/techsupport
Copyright (c) 1986-2020 by Cisco Systems, Inc.
Compiled Thu 30-Jan-20 18:53 by mcpre


Cisco IOS-XE software, Copyright (c) 2005-2020 by cisco Systems, Inc.
All rights reserved.  Certain components of Cisco IOS-XE software are
licensed under the GNU General Public License ("GPL") Version 2.0.  The
software code licensed under GPL Version 2.0 is free software that comes
with ABSOLUTELY NO WARRANTY.  You can redistribute and/or modify such
GPL code under the terms of GPL Version 2.0.  For more details, see the
documentation or "License Notice" file accompanying the IOS-XE software,
or the applicable URL provided on the flyer accompanying the IOS-XE
software.


ROM: IOS-XE ROMMON
BOOTLDR: System Bootstrap, Version 17.1.1[FC2], RELEASE SOFTWARE (P)

abcneras01c95 uptime is 18 weeks, 3 days, 23 hours, 33 minutes
Uptime for this control processor is 18 weeks, 3 days, 23 hours, 34 minutes
System returned to ROM by PowerOn
System restarted at 23:37:40 UTC Wed May 27 2020
System image file is "bootflash:packages.conf"
Last reload reason: PowerOn



This product contains cryptographic features and is subject to United
States and local country laws governing import, export, transfer and
use. Delivery of Cisco cryptographic products does not imply
third-party authority to import, export, distribute or use encryption.
Importers, exporters, distributors and users are responsible for
compliance with U.S. and local country laws. By using this product you
agree to comply with applicable laws and regulations. If you are unable
to comply with U.S. and local laws, return this product immediately.

A summary of U.S. laws governing Cisco cryptographic products may be found at:
http://www.cisco.com/wwl/export/crypto/tool/stqrg.html

If you require further assistance please contact us by sending email to
export@cisco.com.


Technology Package License Information:

------------------------------------------------------------------------------
Technology-package                                     Technology-package
Current                        Type                       Next reboot
------------------------------------------------------------------------------
network-advantage       Smart License                    network-advantage
dna-advantage           Subscription Smart License       dna-advantage          


Smart Licensing Status: REGISTERED/AUTHORIZED

cisco C9500-48Y4C (X86) processor with 1943258K/6147K bytes of memory.
Processor board ID CAT2499L0ZT
23 Virtual Ethernet interfaces
48 TwentyFive Gigabit Ethernet interfaces
4 Hundred Gigabit Ethernet interfaces
32768K bytes of non-volatile configuration memory.
16010532K bytes of physical memory.
11161600K bytes of Bootflash at bootflash:.
1638400K bytes of Crash Files at crashinfo:.
0K bytes of WebUI ODM Files at webui:.

Base Ethernet MAC Address          : ac:3a:67:39:10:e0
Motherboard Assembly Number        : 4873
Motherboard Serial Number          : CAT2499L0ZT
Model Revision Number              : V02
Motherboard Revision Number        : 4
Model Number                       : C9500-48Y4C
System Serial Number               :

Configuration register is 0x102

mgdransfield commented:
ok, updated as suggested

thanks,
Mike

On Wed, Oct 7, 2020 at 10:41 AM Takashi Higashimura <
notifications@github.com> wrote:

> *@tahigash* commented on this pull request.
> ------------------------------
>
> In src/genie/libs/parser/iosxe/c9500/show_platform.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#discussion_r500655359>
> :
>
> > @@ -55,12 +55,14 @@ class ShowVersionSchema(MetaParser):
>                      'model_rev_num': str,
>                      'mb_rev_num': str,
>                      'model_num': str,
> -                    'system_sn': str,
> +                    Optional ('system_sn'): {
>
> thank you for the output. I confirmed no SN in output.
>
> System Serial Number               :
>
> but this doesn't require to change the schema. let's change to Optional
> without having new level.
>
> Optional('system_sn'): str,
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CiscoTestAutomation/genieparser/pull/192#discussion_r500655359>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFM7PB2ZJZ26AOJJFA6XL7LSJOTJJANCNFSM4PRIQNZA>
> .
>

bastell commented:
Hi,
Please resolve the merge conflicts for show interface around helper_address. Otherwise looks good!

mgdransfield commented:
updated to resolve two conflicting files

bastell commented on src/genie/libs/parser/iosxe/show_interface.py:
```diff
@@ -2065,13 +2067,19 @@ def cli(self,interface="",output=None):
+            m = p5_0.match(line)
+            if m:
+                interface_dict[interface]['helper_address'] = \
+                    m.groupdict()['address']
```
`helper_address` must be a list according to the schema. This case should be a list with a single value:
```suggestion
                    [m.groupdict()['address']]
```

bastell reviewed (changes requested):
Please add a changelog according to the template to show which parsers have been updated.

bastell reviewed (approved):
Approved pending fixed merge conflicts
